### PR TITLE
Add Document Service

### PR DESCRIPTION
**Pull request by tjholm**
Also removes deprecated KV service interface.

### CODE REVIEW

File: api/documents/collection_ref.go [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+package documents
+
+import (
+	"fmt"
+
+	v1 "github.com/nitrictech/go-sdk/interfaces/nitric/v1"
+)
+
+// Collection
+type CollectionRef interface {
```
Comment (jyecusch) [inline api/documents/collection_ref.go:24]: Should the interface provide a Name() method, so you can get the name out of the collection (particularly for when the ref wasn't created by the developer directly)?

File: api/documents/collection_ref.go [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+	toWire() *v1.Collection
+}
+
+type collectionRefImpl struct {
+	name           string
+	dc             v1.DocumentServiceClient
+	parentDocument DocumentRef
+}
+
+func (c *collectionRefImpl) Query() Query {
```
Comment (jyecusch) [inline api/documents/collection_ref.go:37]: Doc comments would be useful on these methods

File: api/documents/collection_ref.go [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+
+	if c.GetParent() == nil {
+		return &collectionRefImpl{
+			name: c.GetName(),
+			dc:   dc,
+		}, nil
+	} else {
+		pd, err := documentRefFromWireKey(dc, c.GetParent())
+		if err != nil {
+			return nil, err
```
Comment (jyecusch) [inline api/documents/collection_ref.go:82]: Might be useful to wrap this error to provide context, especially as the depth will be deeper in future.

File: api/documents/document.go [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+
+func (d *documentImpl) Ref() DocumentRef {
+	return d.ref
+}
+
+func (d *documentImpl) Content() map[string]interface{} {
+	return d.content
+}
+
+// Decode -
```
Comment (jyecusch) [inline api/documents/document.go:42]: Doc comment is incomplete

File: api/documents/document_iter.go [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+		return nil, err
+	}
+
+	ref, err := documentRefFromWireKey(i.dc, res.GetDocument().GetKey())
+	if err != nil {
+		return nil, err
+	}
+
+	return &documentImpl{
+		//construct document back reference here...
```
Comment (jyecusch) [inline api/documents/document_iter.go:43]: Is this comment still needed?

File: api/documents/document_ref.go [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+	"context"
+	"fmt"
+
+	"github.com/mitchellh/mapstructure"
+	v1 "github.com/nitrictech/go-sdk/interfaces/nitric/v1"
+	"google.golang.org/protobuf/types/known/structpb"
+)
+
+// DocumentRef - Represents a reference to a document
+type DocumentRef interface {
```
Comment (jyecusch) [inline api/documents/document_ref.go:27]: Should probably have an Id() method to return the Document ID. Primarily for the refs included in get/query responses.

Comment (jyecusch) [reply api/documents/document_ref.go:27]: Also a Parent() method or similar to retrieve the containing CollectionRef

File: api/documents/document_ref.go [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+	Collection(string) (CollectionRef, error)
+}
+
+type documentRefImpl struct {
+	// A shared reference to the top level document service client
+	dc v1.DocumentServiceClient
+	// A reference to this documents collection
+	col CollectionRef
+	// The key for this document
+	key string
```
Comment (jyecusch) [inline api/documents/document_ref.go:50]: This should be id

File: api/documents/document_ref.go [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+	return &collectionRefImpl{
+		name:           c,
+		dc:             d.dc,
+		parentDocument: d,
+	}, nil
+}
+
+func (d *documentRefImpl) Delete() error {
+	_, err := d.dc.Delete(context.TODO(), &v1.DocumentDeleteRequest{
+		Key: &v1.Key{
```
Comment (jyecusch) [inline api/documents/document_ref.go:97]: Should this be using toWireKey() instead?

File: api/documents/document_ref.go [hunk 9]
```diff
@@ -0,0 +1,145 @@
...
+
+func (d *documentRefImpl) Set(content map[string]interface{}) error {
+	sv, err := structpb.NewStruct(content)
+
+	if err != nil {
+		return err
+	}
+
+	_, err = d.dc.Set(context.TODO(), &v1.DocumentSetRequest{
+		Key: &v1.Key{
```
Comment (jyecusch) [inline api/documents/document_ref.go:114]: toWireKey()?

File: api/documents/document_ref.go [hunk 10]
```diff
@@ -0,0 +1,145 @@
...
+}
+
+type DecodeOption interface {
+	Apply(c *mapstructure.DecoderConfig)
+}
+
+// Get -
+func (d *documentRefImpl) Get() (Document, error) {
+	res, err := d.dc.Get(context.TODO(), &v1.DocumentGetRequest{
+		Key: &v1.Key{
```
Comment (jyecusch) [inline api/documents/document_ref.go:131]: toWireKey()?

File: api/documents/document_ref.go [hunk 11]
```diff
@@ -0,0 +1,145 @@
...
+	})
+
+	return err
+}
+
+type DecodeOption interface {
+	Apply(c *mapstructure.DecoderConfig)
+}
+
+// Get -
```
Comment (jyecusch) [inline api/documents/document_ref.go:128]: Doc comment is incomplete

File: api/documents/query.go [hunk 12]
```diff
@@ -0,0 +1,169 @@
...
+import (
+	"context"
+	"fmt"
+
+	v1 "github.com/nitrictech/go-sdk/interfaces/nitric/v1"
+)
+
+// Query - Query interface for Document Service
+type Query interface {
+	// Where - Append an expression to the query
```
Comment (jyecusch) [inline api/documents/query.go:26]: Append *one or more expressions* to the query

File: api/documents/query.go [hunk 13]
```diff
@@ -0,0 +1,169 @@
...
+		return nil, err
+	}
+
+	docs := make([]Document, 0, len(r.GetDocuments()))
+
+	for _, d := range r.GetDocuments() {
+
+		fmt.Println("testing", d.GetKey())
+		// TODO: Handle error
+		// For now we'll just set the back reference to nil
```
Comment (jyecusch) [inline api/documents/query.go:112]: Is this comment out of date?

File: api/documents/query.go [hunk 14]
```diff
@@ -0,0 +1,169 @@
...
+
+	return &FetchResult{
+		Documents:   docs,
+		PagingToken: r.GetPagingToken(),
+	}, nil
+}
+
+func (q *queryImpl) Stream() (DocumentIter, error) {
+	// build the expressions list
+	expressions := make([]*v1.Expression, 0, len(q.exps))
```
Comment (jyecusch) [inline api/documents/query.go:134]: This is repeated code from Fetch, should be extracted

File: api/queues/queue.go [hunk 15]
```diff
@@ -0,0 +1,101 @@
...
+	v1 "github.com/nitrictech/go-sdk/interfaces/nitric/v1"
+)
+
+// Queue - Interface for a Queue reference
+type Queue interface {
+	// Name - The name of the queue
+	Name() string
+	// Send - Push a number of tasks to a queue
+	Send([]*Task) ([]*FailedTask, error)
+	// Recieve - Retrieve tasks from a queue to a maximum of the given depth
```
Comment (jyecusch) [inline api/queues/queue.go:30]: typo in comment

File: api/queues/task.go [hunk 16]
```diff
@@ -0,0 +1,94 @@
...
+type Task struct {
+	// ID - Unique ID for the task
+	ID string
+	// PayloadType - Deserialization hint for interprocess communication
+	PayloadType string
+	// Payload - The payload to include in this task
+	Payload map[string]interface{}
+}
+
+type RecievedTask interface {
```
Comment (jyecusch) [inline api/queues/task.go:34]: typo in interface name

**Review by jyecusch — CHANGES_REQUESTED** [review]

File: api/documents/collection_ref.go (hunk 1 above)
Comment (tjholm) [reply api/documents/collection_ref.go:24]: Done

File: api/documents/collection_ref.go (hunk 2 above)
Comment (tjholm) [reply api/documents/collection_ref.go:37]: done

File: api/documents/collection_ref.go (hunk 3 above)
Comment (tjholm) [reply api/documents/collection_ref.go:82]: The error highlighted in this line isn't related to the collection depth, but I've crafted an explicit error for the collection depth exceed error.

File: api/documents/document.go (hunk 4 above)
Comment (tjholm) [reply api/documents/document.go:42]: Done

File: api/documents/document_iter.go (hunk 5 above)
Comment (tjholm) [reply api/documents/document_iter.go:43]: Done

File: api/documents/document_ref.go (hunk 6 above)
Comment (tjholm) [reply api/documents/document_ref.go:27]: Done

File: api/documents/document_ref.go (hunk 7 above)
Comment (tjholm) [reply api/documents/document_ref.go:50]: done

File: api/documents/document_ref.go (hunk 8 above)
Comment (tjholm) [reply api/documents/document_ref.go:97]: done

File: api/documents/document_ref.go (hunk 9 above)
Comment (tjholm) [reply api/documents/document_ref.go:114]: done

File: api/documents/document_ref.go (hunk 10 above)
Comment (tjholm) [reply api/documents/document_ref.go:131]: done

File: api/documents/document_ref.go (hunk 11 above)
Comment (tjholm) [reply api/documents/document_ref.go:128]: done

File: api/documents/query.go (hunk 12 above)
Comment (tjholm) [reply api/documents/query.go:26]: done

File: api/documents/query.go (hunk 13 above)
Comment (tjholm) [reply api/documents/query.go:112]: done

File: api/documents/query.go (hunk 14 above)
Comment (tjholm) [reply api/documents/query.go:134]: done

File: api/queues/queue.go (hunk 15 above)
Comment (tjholm) [reply api/queues/queue.go:30]: done.

File: api/queues/task.go (hunk 16 above)
Comment (tjholm) [reply api/queues/task.go:34]: done.

**Review by jyecusch — APPROVED** [review]

**Pull request merged**